### PR TITLE
Autograd functional API should enable_grad

### LINKED ISSUE #44640: autograd functional APIs should enable_grad
## 🐛 Bug

Currently the functional API will return zero gradients if it's wrapped in a no_grad, and without issuing any error messages. This is unexpected from the user end because we only use these methods when we want gradients.

I was using this inside custom Functions, where the default is no_grad. There might be other scenarios where a user might accidentally call these inside a no_grad.

## To Reproduce

Reproducing code:

```
import torch
from torch.autograd.functional import vjp


def exp_reducer(x):
    return x.exp().sum(dim=1)


inputs = torch.rand(4, 4)
v = torch.ones(4)

with torch.no_grad():
    print(vjp(exp_reducer, inputs, v))

with torch.enable_grad():
    print(vjp(exp_reducer, inputs, v))
```

stdout:
```
(tensor([6.9459, 7.6220, 6.1336, 6.0577]), tensor([[0., 0., 0., 0.],
        [0., 0., 0., 0.],
        [0., 0., 0., 0.],
        [0., 0., 0., 0.]]))
(tensor([6.9459, 7.6220, 6.1336, 6.0577]), tensor([[2.2766, 2.3207, 1.3200, 1.0287],
        [2.7153, 1.4122, 1.6791, 1.8154],
        [1.5301, 1.7467, 1.7776, 1.0792],
        [1.2966, 1.0290, 1.8711, 1.8610]]))
```

## Expected behavior

vjp correctly returned even inside no_grad(), or a clear error message.

## Environment

PyTorch version: 1.6.0

cc @ezyang @SsnL @albanD @zou3519 @gqchen @pearu

### PR DESCRIPTION
**Pull request by iramazanli**
Fixes #44640

### CODE REVIEW

**Comment by iramazanli** [comment]
More efficient implementation creating graphs only when its necessary

File: torch/autograd/functional.py [hunk 1]
```diff
@@ -245,28 +245,33 @@ def vjp(func, inputs, v=None, create_graph=False, strict=False):
...
+            _, v = _as_tuple(v, "v", "vjp")
+            v = _grad_preprocess(v, create_graph=create_graph, need_graph=False )
+            _validate_v(v, outputs, is_outputs_tuple)
+        else:
+            if len(outputs) != 1 or outputs[0].nelement() != 1:
+                raise RuntimeError("The vector v can only be None if the "
+                                   "user-provided function returns "
+                                   "a single Tensor with a single element.")
+    
+    if create_graph:
```
Comment (albanD) [inline torch/autograd/functional.py:267]: To avoid code duplication you may want to do something like:
```python
enable_grad = True if create_graph else torch.is_grad_enabled()
with torch.set_grad_enabled(enable_grad):
      grad_res = _autograd_grad(outputs, inputs, v, create_graph=create_graph)
      vjp = _fill_in_zeros(grad_res, inputs, strict, create_graph, "back")
```

File: torch/autograd/functional.py [hunk 2]
```diff
@@ -329,34 +334,39 @@ def jvp(func, inputs, v=None, create_graph=False, strict=False):
         for forward mode AD in PyTorch at the moment.
     """
 
-    is_inputs_tuple, inputs = _as_tuple(inputs, "inputs", "jvp")
-    inputs = _grad_preprocess(inputs, create_graph=create_graph, need_graph=True)
+    with torch.enable_grad():   
+        is_inputs_tuple, inputs = _as_tuple(inputs, "inputs", "jvp")
+        inputs = _grad_preprocess(inputs, create_graph=create_graph, need_graph=True)
```
Comment (albanD) [inline torch/autograd/functional.py:339]: Does the grad preprocess actually needs to always run with grad mode enabled? Same question for the other functions

File: torch/autograd/functional.py [hunk 3]
```diff
@@ -245,28 +245,33 @@ def vjp(func, inputs, v=None, create_graph=False, strict=False):
...
+        is_inputs_tuple, inputs = _as_tuple(inputs, "inputs", "vjp")
+        inputs = _grad_preprocess(inputs, create_graph=create_graph, need_graph=True)
+
+        outputs = func(*inputs)
+        is_outputs_tuple, outputs = _as_tuple(outputs, "outputs of the user-provided function", "vjp")
+        _check_requires_grad(outputs, "outputs", strict=strict)
+
+        if v is not None:
+            _, v = _as_tuple(v, "v", "vjp")
+            v = _grad_preprocess(v, create_graph=create_graph, need_graph=False )
```
Comment (albanD) [inline torch/autograd/functional.py:259]: Does this actually always needs grad to be enabled?

**Review by albanD — COMMENTED** [review]
I don't think you wanted to commit the `.nojekyll` file right?

**Review by iramazanli — COMMENTED** [review]
yes I don't want :) so i removed the file

**Review by albanD — COMMENTED** [review]
LGTM 
You need to fix the lint though
And you can rebase on top of the `viable/strict` branch to make sure you don't get any CI failures due to a broken master.

**Review by albanD — APPROVED** [review]
LGTM thanks for removing the jekyll file.

**Comment by albanD** [comment]
btw the jekyll file is back in the PR and should be removed before merging.

File: torch/autograd/functional.py [hunk 4]
```diff
@@ -250,26 +250,31 @@ def vjp(func, inputs, v=None, create_graph=False, strict=False):
...
+            _, v = _as_tuple(v, "v", "vjp")
+            v = _grad_preprocess(v, create_graph=create_graph, need_graph=False)
+            _validate_v(v, outputs, is_outputs_tuple)
+        else:
+            if len(outputs) != 1 or outputs[0].nelement() != 1:
+                raise RuntimeError("The vector v can only be None if the "
+                                   "user-provided function returns "
+                                   "a single Tensor with a single element.")
+
+    if create_graph:
```
Comment (albanD) [inline torch/autograd/functional.py:271]: I think you missed the proposed update here no?
In particular, you can 
```python
with torch.enable_grad(create_graph):
        grad_res = _autograd_grad(outputs, inputs, v, create_graph=create_graph)
        vjp = _fill_in_zeros(grad_res, inputs, strict, create_graph, "back")
```

I think you can do this update for all functions below.

**Review by albanD — APPROVED** [review]
Just a small update to reduce code duplication, good to go otherwise.

Thanks for the rebase.

File: torch/autograd/functional.py (hunk 4 above)
Comment (iramazanli) [reply torch/autograd/functional.py]: Thanks for reminding :)
